dask/dataframe/io/parquet/core.py

```diff
@@ -0,0 +1,452 @@
+        path,
+        columns,
+        filters,
+        categories,
+        index,
+        storage_options,
+        engine,
+        gather_statistics,
+    )
+
+    if isinstance(engine, str):
+        engine = get_engine(engine)
+
+    fs, fs_token, paths = get_fs_token_paths(
+        path, mode="rb", storage_options=storage_options
+    )
+
+    paths = sorted(paths, key=natural_sort_key)  # numeric rather than glob ordering
+
+    meta, statistics, parts = engine.read_metadata(
+        fs,
+        fs_token,
+        paths,
+        categories=categories,
+        index=index,
+        gather_statistics=gather_statistics,
+    )
+
+    if columns is None:
+        columns = meta.columns
+
+    if not set(columns).issubset(set(meta.columns)):
+        raise KeyError(
+            "The following columns were not found in the dataset %s\n"
+            "The following columns were found %s"
+            % (set(columns) - set(meta.columns), meta.columns)
+        )
+
+    if statistics:
+        parts, statistics = zip(
+            *[
+                (part, stats)
+                for part, stats in zip(parts, statistics)
+                if stats["num-rows"] > 0
+            ]
+        )
+
+        if filters:
+            parts, statistics = apply_filters(parts, statistics, filters)
+
+        out = sorted_columns(statistics)
+
+        if index and out:
+            out = [o for o in out if o["name"] == index]  # only one valid column
+        if index is not False and len(out) == 1:
+            divisions = out[0]["divisions"]
+            index = out[0]["name"]
+        elif index is not False and len(out) > 1:
+            if any(o["name"] == "index" for o in out):
+                [o] = [o for o in out if o["name"] == "index"]
+                divisions = o["divisions"]
+                index = o["name"]
+            else:
+                warnings.warn(
+                    "Multiple sorted columns found: "
+                    + ", ".join(o["name"] for o in out)
+                )
+                divisions = [None] * (len(parts) + 1)
+        else:
+            divisions = [None] * (len(parts) + 1)
+    else:
+        divisions = [None] * (len(parts) + 1)
+
+    if index and index not in columns:
+        columns = list(columns) + [index]
+
+    meta = meta[columns]
+
+    subgraph = {
+        (name, i): (
+            read_parquet_part,
+            engine.read_partition,
+            fs,
+            meta,
+            part["piece"],
+            columns,
+            index,
+            part["kwargs"],
+        )
+        for i, part in enumerate(parts)
+    }
+    if index:
+        meta = meta.set_index(index)
+
+    return new_dd_object(subgraph, name, meta, divisions)
+
+
+def read_parquet_part(func, fs, meta, part, columns, index, kwargs):
+    """ Read a part of a parquet dataset """
+    df = func(fs, part, columns, **kwargs)
+    if not len(df):
+        df = meta
+    if index:
+        df = df.set_index(index)
```

Comment:
I think this should happen within the `func`.

Reply:
My thought was that I didn't want the engine to have to think about this.  Parquet doesn't have a sense of Pandas index, that's on the Pandas/Dask side.  We just want a pile of columns from the Parquet reader, we'll then coerce it into the form that we think best.  This again falls into the "lets not ask much from the readers" philosophy

Reply:
> We just want a pile of columns from the Parquet reader

unfortunately, this is no longer true with the rangeindexes - the pieces will have an index set that is not a column, making things more complicated

Reply:
Well, if we're roundtripping our own data then this won't matter if we also don't write the index objects (but rather elevate indexes to columns prior to writing).  

However I suppose that some other system might write lots of pandas objects that will have range indexes that we want to capture.  In this case we would be blind if we also identified a sorted column and set that as the index.  If this is the only case where this occurs, then I'm not particularly concerned.  I might be missing things here though.